src/Microsoft.AspNetCore.Routing/Matching/DefaultEndpointSelector.cs

```diff
@@ -24,16 +24,30 @@ public DefaultEndpointSelector(IEnumerable<MatcherPolicy> matcherPolicies)
             _selectorPolicies = matcherPolicies.OrderBy(p => p.Order).OfType<IEndpointSelectorPolicy>().ToArray();
         }
 
-        public override Task SelectAsync(
+        public override async Task SelectAsync(
             HttpContext httpContext,
             EndpointFeature feature,
             CandidateSet candidateSet)
         {
+            var selectorPolicies = _selectorPolicies;
             for (var i = 0; i < _selectorPolicies.Length; i++)
             {
-                _selectorPolicies[i].Apply(httpContext, candidateSet);
+                await selectorPolicies[i].ApplyAsync(httpContext, feature, candidateSet);
```

Comment:
What is the perf hit of this? If it is called once per-match then state machine overhead will be tiny but I assume there is a Task allocation. Is it worth providing a non-async path?

Reply:
Ok! I couldn't remember exactly when a Task was allocated. Optimizing a state machine called once per match is def not worth it.